src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,934 @@
+         | _ -> false
+        )
+    in  
+    if env_comp = -2 || env_comp > 0 then false else
+    if is_bot_env t1 || is_top t2 then true else
+    if is_bot_env t2 || is_top t1 then false else (
+      let m1, m2 = Option.get t1.d, Option.get t2.d in
+      let m1' = if env_comp = 0 then m1 else dim_add (Environment.dimchange t1.env t2.env) m1 in
+      Array.fold_lefti (fun b i t -> b && implies m1' t i) true m2)
+
+  let leq a b = timing_wrap "leq" (leq a) b
+
+  let leq t1 t2 =
+    let res = leq t1 t2 in
+    if M.tracing then M.tracel "leq" "leq a: %s b: %s -> %b \n" (show t1) (show t2) res ;
+    res
+
+  let join a b =     
+    let join_d ad bd = 
+      (*This is the table which is later grouped*)
+      let table = BatList.map2i (fun i a b -> (i,a,b)) (Array.to_list ad) (Array.to_list bd) in
+      let const_offset t = Tuple2.second t in
+      let diff t1 t2 = Z.((const_offset t1) - (const_offset t2)) in
+      (*compare two variables for grouping depending on delta function and reference index*)
+      let cmp_z (_, t1i, t2i) (_, t1j, t2j) = 
+        let cmp_z_ref (x,_) (y,_): int =
+          match x, y with
+          | None, None -> 0
+          | None, Some _ -> -1
+          | Some _, None -> 1
+          | Some ii, Some ij -> ii - ij 
+        in
+        let diff_e1 = cmp_z_ref t1i t1j in
+        if diff_e1 <> 0 then diff_e1 
+        else
+          let diff_e2 = cmp_z_ref t2i t2j in
+          if diff_e2 <> 0 then diff_e2 else 
+            Z.to_int (Z.((diff t1i t2i) - (diff t1j t2j)))
+      in
+      (*Calculate new components as groups*)
+      let new_components = BatList.group cmp_z table in
+    (*Adjust the domain array to represent the new components*)
+      let modify idx_h b_h (idx, (opt1, z1), (opt2, z2)) =
+        if idx_h = idx then ad.(idx) <- (Some idx, Z.zero)
+        else if Option.(opt1 = opt2) && Z.(z1 = z2) then ()
+        else ad.(idx) <- (Some idx_h, Z.(z1 - b_h))
+      in
+      let iterate l =
+        match l with 
+        | (idx_h, (_, b_h), _) :: t -> List.iter (modify idx_h b_h) l
+        | [] -> () (*This should not happen, consider throughing exception*)       
+    in
+    List.iter iterate new_components; Some ad in
+    (*Normalize the two domains a and b such that both talk about the same variables*)
+    if is_bot_env a then b else if is_bot_env b then a 
+    else
+      match Option.get a.d, Option.get b.d with
+      | x, y when is_top a || is_top b -> let new_env = Environment.lce a.env b.env 
+        in (identity new_env)
+      | x, y when (Environment.compare a.env b.env <> 0) ->
+        let sup_env = Environment.lce a.env b.env in
+        let mod_x = dim_add (Environment.dimchange a.env sup_env) x in
+        let mod_y = dim_add (Environment.dimchange b.env sup_env) y in
+        {d = join_d mod_x mod_y; env = sup_env}
+      | x, y when EArray.equal x y -> {d = Some x; env = a.env}
+      | x, y  -> {d = join_d x y; env = a.env}
```

Comment:
The join is definitely a move in the right direction! It looks like the join currently modifies its first argument though in case the environments agree, nothing is top, and they are not identical. (I could be mistaken).

If it does that is a problem, we should return a new array here.

Reply:
I did it to save some memory and instructions but if you wanna have a new array I can also use a copy.

Reply:
It's ok (and in fact desirable) to modify the argument in functions that are called **`_with`** to save memory and time as you suggested as the caller expects this call to potentially change the arguments.

For all other functions, it is incorrect to modify the argument, as the caller relies on the arguments remaining unchanged. Imagine you have a join point in the control flow graph and want to compute the information after the join

```
if (...) {
  // Abstract state  A1
} else {
  // Abstract state  A2
}

// Abstract state: A'
```

`A' = join A1 A2`. Now, if you modify the argument, you have silently made the information we have for `A1` more imprecise, as it is now the same as `A'`